### ISSUE TITLE
Segfault in opal_generic_simple_pack when sending struct with a >4GB block

### ISSUE THREAD

**Issue by psalz** [open-mpi/ompi#7147]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

Tested with v4.0.0 and v4.0.2

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Tarball from open-mpi.org. No additional parameters provided to `configure`.

### Please describe the system on which you are running

* Operating system/version: Ubuntu 18.04 LTS
* Computer hardware: AMD Ryzen Threadripper 2920X, 32GB RAM
* Network type: Vader/TCP (10GE)

-----------------------------

I'm running into a segfault inside `opal_generic_simple_pack` when trying to send a custom struct that contains two members, with one block that is larger than (or equal to) `UINT_MAX` (i.e., 4 GB). I'm not sure whether this is related to the known issue mentioned in the 4.0.2 changelog, but since I'm getting a plain segfault (as opposed to some error message), I believe this is a different corner case.

Here's a minimal example:

```cpp
#include <cassert>
#include <cstdio>
#include <memory>
#include <cstdlib>
#include <vector>

#include <mpi.h>

//constexpr size_t BUFFER_SIZE = 2ull << 30; // Works
constexpr size_t BUFFER_SIZE = 2ull << 31; // Breaks

constexpr int BLOCK_SIZE = 1024;

struct Header {
    int foo;
    int bar;
};

int main(int argc, char* argv[]) {
    assert(BUFFER_SIZE % BLOCK_SIZE == 0); 

    MPI_Init(&argc, &argv);
    int size;
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    if(size != 2) return EXIT_FAILURE;
    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    MPI_Datatype block_type;
    MPI_Type_contiguous(BLOCK_SIZE, MPI_BYTE, &block_type);
    MPI_Type_commit(&block_type);

    std::vector<char> buffer(BUFFER_SIZE);

    Header header;

    MPI_Datatype struct_type;
    // Remove the header block to make it work again
    std::vector<int> block_lengths = { 
        sizeof(Header), // Remove
        BUFFER_SIZE / BLOCK_SIZE};
    std::vector<MPI_Aint> block_disps = { 
        reinterpret_cast<MPI_Aint>(&header), // Remove
        reinterpret_cast<MPI_Aint>(buffer.data())};
    std::vector<MPI_Datatype> block_types = { 
        MPI_BYTE, // Remove
        block_type};
    MPI_Type_create_struct(block_lengths.size(), block_lengths.data(),
        block_disps.data(), block_types.data(), &struct_type);
    MPI_Type_commit(&struct_type);

    if(rank == 1) {
        header.foo = 3;
        header.bar = 42; 
        MPI_Send(MPI_BOTTOM, 1, struct_type, 0, 0, MPI_COMM_WORLD);
    } else {
        MPI_Recv(MPI_BOTTOM, 1, struct_type, 1, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        printf("Received header foo %d bar %d\n", header.foo, header.bar);
    }   

    MPI_Type_free(&struct_type);
    MPI_Type_free(&block_type);
    MPI_Finalize();
    return EXIT_SUCCESS;
}
```

I tested this both using shared memory (Vader) on a single node, as well as using TCP on a single node and two nodes. All variants cause the segfault.

What's interesting is that this only happens if the header is included in the struct. If you remove the lines indicated above, the `BUFFER_SIZE` can be larger than `UINT_MAX` and everything still works!

**Comment by bosilca** [comment]
There is an unfortunately multiplication between 2 uint32_t that nullify with the length of the datatype in this example. I have a patch but I need to make sure we cope with this case in all other instances.

**Comment by jsquyres** [comment]
This fix will be included in the upcoming v4.0.x release.

It will _not_ be included in upcoming v3.0.x or v3.1.x releases; back-porting the fixes is unfortunately out of scope.

**Comment by psalz** [comment]
That's great news! Thanks for handling this so quickly!
